api/include/opentelemetry/logs/logger.h

```diff
@@ -55,65 +54,322 @@ class Logger
+    // Create a LogRecord to hold this information
+    LogRecord r;
+    r.severity = severity;
+    r.body     = message;
+
+    // Call the main Log(LogRecord) method
+    Log(r);
   }
 
-  inline void log(Severity severity,
-                  nostd::string_view message,
-                  core::SystemTimestamp time) noexcept
+  /**
+   * Writes a log.
+   * @param severity The log's severity
+   * @param name The name of the log
+   * @param message The message to log
+   */
+  inline void Log(Severity severity, nostd::string_view name, nostd::string_view message) noexcept
   {
-    // creates a LogRecord object with given parameters, then calls log(LogRecord)
+    // Create a LogRecord to hold this information
     LogRecord r;
-    r.severity  = severity;
-    r.body      = message;
-    r.timestamp = time;
+    r.severity = severity;
+    r.name     = name;
+    r.body     = message;
 
-    log(r);
+    // Call the main Log(LogRecord) method
+    Log(r);
   }
 
-  /** Overloaded methods for structured logging**/
-  // TODO: separate this method into separate methods since it is not useful for user to create
-  // empty logs
+  /**
+   * Writes a log.
+   * @param severity The severity of the log, from 1 to 24
+   * @param attributes A key/value object
+   */
   template <class T,
             nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr>
-  inline void log(Severity severity       = Severity::kDefault,
-                  nostd::string_view name = "",
-                  const T &attributes     = {}) noexcept
+  inline void Log(Severity severity, const T &attributes) noexcept
   {
-    log(severity, name, common::KeyValueIterableView<T>(attributes));
+    // Create a LogRecord to hold this information
+    LogRecord r;
+    r.severity   = severity;
+    r.attributes = nostd::shared_ptr<common::KeyValueIterable>(
+        new common::KeyValueIterableView<T>{attributes});
+
+    // Call the main Log(LogRecord) method
+    Log(r);
   }
 
-  inline void log(Severity severity,
-                  nostd::string_view name,
-                  const common::KeyValueIterable &attributes) noexcept
+  /**
+   * Writes a log.
+   * @param severity The severity of the log, from 1 to 24
+   * @param name The name of the log
+   * @param attributes A key/value object
+   */
+  template <class T,
+            nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr>
+  inline void Log(Severity severity, nostd::string_view name, const T &attributes) noexcept
   {
-    // creates a LogRecord object with given parameters, then calls log(LogRecord)
+    // Create a LogRecord to hold this information
     LogRecord r;
     r.severity   = severity;
     r.name       = name;
-    r.attributes = attributes;
+    r.attributes = nostd::shared_ptr<common::KeyValueIterable>(
+        new common::KeyValueIterableView<T>{attributes});
+
+    // Call the main Log(LogRecord) method
+    Log(r);
+  }
```

Comment:
It would be important to support the `log` function to take the attributes as part of function call.
Eg, 
```
logger.log(severtity.kTrace,  "test", {{"k1", "v1"}, {"k2","v2"}}
```
Refer to trace api ([Tracer::StartSpan](https://github.com/open-telemetry/opentelemetry-cpp/blob/11cc6cd8b18932c92e19bc92eb87aff1c7c5baad/api/include/opentelemetry/trace/tracer.h#L70) function ) on how this can be achieved.

Reply:
Yup that looks pretty handy for the user, I'll add that in!

Reply:
Look the way we do in Trace::StartSpan overload. `initializer_list` overloded function calls `nostd::span` overloaded function which calls `const T*` overloaded function

Reply:
Yea I took at look at that, but with traces the `initializer_list` immediately gets stored in memory in a recordable inside the SDK (after being converted to `nostd::span` -> `const T*`). In our case though, only the KeyValueIterable gets stored in the LogRecord, but the KeyValueIterable doesn't own the attributes/resources but just points to it, so the `initializer_list` has no memory to live at.